pkg/ebpftracer/c/tracee.bpf.c

```diff
@@ -6092,125 +5693,52 @@ statfunc u32 cgroup_skb_generic(struct __sk_buff *ctx, void *cgrpctxmap)
-
-    u32 ihl = 0;
-    switch (family) {
-        case PF_INET:
-            if (nethdrs->iphdrs.iphdr.version != 4) // IPv4
-                return 1;
-
-            ihl = nethdrs->iphdrs.iphdr.ihl;
-            if (ihl > 5) { // re-read IPv4 header if needed
-                size -= get_type_size(struct iphdr);
-                size += ihl * 4;
-                bpf_skb_load_bytes_relative(ctx, 0, dest, size, 1);
-            }
-
-            switch (nethdrs->iphdrs.iphdr.protocol) {
-                case IPPROTO_TCP:
-                case IPPROTO_UDP:
-                case IPPROTO_ICMP:
-                    break;
-                default:
-                    return 1; // unsupported proto
-            }
-
-            // add IPv4 header items to indexer
-            indexer.ip_csum = nethdrs->iphdrs.iphdr.check;
-            indexer.src.in6_u.u6_addr32[0] = nethdrs->iphdrs.iphdr.saddr;
-            indexer.dst.in6_u.u6_addr32[0] = nethdrs->iphdrs.iphdr.daddr;
-            break;
-
-        case PF_INET6:
-            // TODO: dual-stack IP implementation unsupported for now
-            // https://en.wikipedia.org/wiki/IPv6_transition_mechanism
-            if (nethdrs->iphdrs.ipv6hdr.version != 6) // IPv6
-                return 1;
-
-            switch (nethdrs->iphdrs.ipv6hdr.nexthdr) {
-                case IPPROTO_TCP:
-                case IPPROTO_UDP:
-                case IPPROTO_ICMPV6:
-                    break;
-                default:
-                    return 1; // unsupported proto
-            }
-
-            // add IPv6 header items to indexer
-            __builtin_memcpy(&indexer.src.in6_u, &nethdrs->iphdrs.ipv6hdr.saddr.in6_u, 4 * sizeof(u32));
-            __builtin_memcpy(&indexer.dst.in6_u, &nethdrs->iphdrs.ipv6hdr.daddr.in6_u, 4 * sizeof(u32));
-            break;
-
-        default:
-            return 1; // verifier
-    }
-
-    net_event_context_t *neteventctx;
-    neteventctx = bpf_map_lookup_elem(cgrpctxmap, &indexer); // obtain event context
-    if (!neteventctx) {
+    u64 cookie = bpf_get_socket_cookie(ctx);
+    net_task_context_t *netctx = bpf_map_lookup_elem(&net_taskctx_map, &cookie); // obtain event context
+    if (!netctx) {
         // 1. kthreads receiving ICMP and ICMPv6 (e.g dest unreach)
-        // 2. tasks not being traced
-        // 3. unknown (yet) sockets (need egress packet to link task and inode)
+        // 2. tasks not being traced (socket was created before agent started)
         // ...
         return 1;
     }
 
     // Skip if cgroup is muted.
-    u64 cgroup_id = neteventctx->eventctx.task.cgroup_id;
+    u64 cgroup_id = netctx->taskctx.cgroup_id;
     if (bpf_map_lookup_elem(&ignored_cgroups_map, &cgroup_id)) {
         return 1;
     }
 
-    neteventctx->md.header_size = size; // add header size to offset
-
-    u32 ret = CGROUP_SKB_HANDLE(proto);
+    int zero = 0;
+    net_event_context_t *neteventctx = bpf_map_lookup_elem(&cgroup_skb_events_scratch_map, &zero);
```

Comment:
It would be nice, if we could use our new next free scratch buffer helper. Even though it works on ’event_context_t’, it shouldn't be a problem, as the net context embeds it as the first field.